pkg/sql/virtual_table.go

```diff
@@ -24,20 +25,79 @@ import (
 // nil). If there is an error, then return (nil, error).
 type virtualTableGenerator func() (tree.Datums, error)
 
+type virtualTableGeneratorResponse struct {
+	datums tree.Datums
+	err    error
+}
+
+var errRowsDone = errors.New("predefined error that the generator has been canceled")
+
+// setupGenerator takes in a worker that generates rows eagerly and transforms
+// it into a lazy row generator. It returns three functions:
+// * cleanup: Performs all cleanup. May only be called once.
+// * startGenerator: The generator version of the input worker. Intended to
+//   be started as a goroutine.
+// * next: A handle that can be called to generate a row from the worker. Next
+//   cannot be called once cleanup has been called.
+func setupGenerator(
+	worker func(addRow func(...tree.Datum) error) error,
+) (cleanup func(), startGenerator func(), next virtualTableGenerator) {
+	done, send := make(chan struct{}), make(chan virtualTableGeneratorResponse)
+	cleanup = func() {
+		close(done)
+	}
+	addRow := func(datums ...tree.Datum) error {
+		// We need another select here so that in case cleanup has
+		// been before we started waiting in addRow.
+		select {
+		case <-done:
+			return errRowsDone
+		default:
+		}
+
+		select {
+		case <-done:
+			return errRowsDone
+		case send <- virtualTableGeneratorResponse{datums: datums}:
+		}
+		return nil
+	}
+
+	startGenerator = func() {
+		err := worker(addRow)
+		if err == errRowsDone {
+			return
+		}
+		// Notify that we are done sending rows. This is safe because there should
+		// be someone reading from next(). There would only be no one reading from
+		// next() is cleanup() was called, which means that err would be equal to
+		// errRowsDone.
+		send <- virtualTableGeneratorResponse{err: err}
```

Comment:
can we close the send channel instead to signify completion?

Reply:
done